service/domain/blobs/replication/wants.go

```diff
@@ -0,0 +1,218 @@
+package replication
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/boreq/errors"
+	"github.com/planetary-social/go-ssb/logging"
+	"github.com/planetary-social/go-ssb/service/domain/messages"
+	"github.com/planetary-social/go-ssb/service/domain/refs"
+	"github.com/planetary-social/go-ssb/service/domain/transport"
+)
+
+type WantsProcess struct {
+	incomingLock sync.Mutex
+	incoming     []incomingStream
+
+	remoteWantsLock sync.Mutex
+	remoteWants     map[string]struct{}
+
+	wantListStorage WantListStorage
+	blobStorage     BlobSizeRepository
+	downloader      Downloader
+	logger          logging.Logger
+}
+
+func NewWantsProcess(
+	wantListStorage WantListStorage,
+	blobStorage BlobSizeRepository,
+	downloader Downloader,
+	logger logging.Logger,
+) *WantsProcess {
+	return &WantsProcess{
+		remoteWants: make(map[string]struct{}),
+
+		wantListStorage: wantListStorage,
+		blobStorage:     blobStorage,
+		downloader:      downloader,
+		logger:          logger.New("wants_process"),
+	}
+}
+
+func (p *WantsProcess) AddIncoming(ctx context.Context, ch chan<- messages.BlobWithSizeOrWantDistance) {
+	p.logger.Debug("adding incoming")
+
+	p.incomingLock.Lock()
+	defer p.incomingLock.Unlock()
+
+	stream := newIncomingStream(ctx, ch)
+	p.incoming = append(p.incoming, stream)
+
+	go func() {
+		defer close(ch)
+		defer func() {
+			if err := p.cleanupIncomingStream(stream); err != nil {
+				p.logger.WithError(err).Error("could not clean up a stream")
+			}
+		}()
+		p.incomingLoop(stream.ctx, stream.ch)
+	}()
+}
+
+func (p *WantsProcess) AddOutgoing(ctx context.Context, ch <-chan messages.BlobWithSizeOrWantDistance, peer transport.Peer) {
+	p.logger.Debug("adding outgoing")
+	go p.outgoingLoop(ctx, ch, peer)
+}
+
+func (p *WantsProcess) incomingLoop(ctx context.Context, ch chan<- messages.BlobWithSizeOrWantDistance) {
+	if err := p.respondToPreviousWants(); err != nil {
+		p.logger.WithError(err).Error("failed to respond to previous wants")
+	}
+
+	for {
+		wl, err := p.wantListStorage.GetWantList()
+		if err != nil {
+			p.logger.WithError(err).Error("could not get the want list")
+			continue
+		}
+
+		for _, v := range wl.List() {
```

Comment:
I think that we are missing dropping a blob from the wants list once we fetch it. That means we will always tell all our peers about all of our wants what is not optimal I guess. Should we add this to a todo list in https://github.com/planetary-social/go-ssb/issues/1?

Reply:
Correct.

Reply:
I added this to #1. The want list has to be redone in general, it is completely temporary.